src/AbstractSQLCompiler.ts

```diff
@@ -15,6 +15,211 @@ import sbvrTypes = require('@resin/sbvr-types');
+	13?: Y;
+	14?: Y;
+	15?: Y;
+	16?: Y;
+	17?: Y;
+	18?: Y;
+	19?: Y;
+	20?: Y;
+	21?: Y;
+}
+
+export type BooleanNode = ['Boolean', boolean];
+export interface EqualsNode extends TwoArgNodeType<'Equals', AbstractSqlType> {}
+export interface NotEqualsNode
+	extends TwoArgNodeType<'NotEquals', AbstractSqlType> {}
+export interface GreaterThanNode
+	extends TwoArgNodeType<'GreaterThan', AbstractSqlType> {}
+export interface GreaterThanOrEqualNode
+	extends TwoArgNodeType<'GreaterThanOrEqual', AbstractSqlType> {}
+export interface LessThanNode
+	extends TwoArgNodeType<'LessThan', AbstractSqlType> {}
+export interface LessThanOrEqualNode
+	extends TwoArgNodeType<'LessThanOrEqual', AbstractSqlType> {}
+export interface InNode
+	extends TwoVarArgNodeType<
+		'In',
+		FieldNode | ReferencedFieldNode,
+		AbstractSqlType
+	> {}
+export interface ExistsNode extends OneArgNodeType<'Exists', AbstractSqlType> {}
+export interface NotNode extends OneArgNodeType<'Not', BooleanTypeNodes> {}
+export interface AndNode extends VarArgNodeType<'And', BooleanTypeNodes> {}
+export interface OrNode extends VarArgNodeType<'Or', BooleanTypeNodes> {}
+export type BooleanTypeNodes =
+	| BooleanNode
+	| EqualsNode
+	| NotEqualsNode
+	| GreaterThanNode
+	| GreaterThanOrEqualNode
+	| LessThanNode
+	| LessThanOrEqualNode
+	| InNode
+	| ExistsNode
+	| NotNode
+	| AndNode
+	| OrNode
+	| UnknownTypeNodes;
+
+export type NumberNode = ['Number', Number];
+export type CountNode = ['Count', '*'];
+export type NumberTypeNodes = NumberNode | CountNode | UnknownTypeNodes;
+
+export type FieldNode = ['Field', string];
+export type ReferencedFieldNode = ['ReferencedField', string, string];
+export type BindNode = ['Bind', string, string?];
+export interface CastNode
+	extends TwoVarArgNodeType<'Cast', AbstractSqlType, string> {}
+export type UnknownTypeNodes =
+	| FieldNode
+	| ReferencedFieldNode
+	| BindNode
+	| CastNode
+	| AbstractSqlQuery;
+
+export type TextNode = ['Text', string];
+export interface ConcatenateNode
+	extends VarArgNodeType<'Concatenate', TextTypeNodes> {}
+export type LikeNode = ['Like', '*'];
+export interface ReplaceNode
+	extends ThreeArgNodeType<'Replace', TextTypeNodes> {}
+export type TextTypeNodes =
+	| ConcatenateNode
+	| LikeNode
+	| ReplaceNode
+	| UnknownTypeNodes;
+
+export type SelectQueryNode = [
+	'SelectQuery',
+	...Array<
+		SelectNode | FromNode | WhereNode | OrderByNode | LimitNode | OffsetNode
+	>
+];
+export type UnionQueryNode = ['UnionQuery', ...Array<SelectQueryNode>];
+
+export interface SelectNode
+	extends OneArgNodeType<'Select', AbstractSqlType[]> {}
+export interface FromNode
+	extends TwoArgNodeType<
+		'From',
+		| SelectQueryNode
+		| UnionQueryNode
+		| TableNode
+		| AliasNode<SelectQueryNode | UnionQueryNode | TableNode>
+	> {}
+export type TableNode = ['Table', string];
+export type WhereNode = ['Where', BooleanTypeNodes];
+// export type GroupByNode = ['GroupBy', Array<['ASC' | 'DESC', FieldNode | ReferencedFieldNode]>];
```

Comment:
Is that commented out by accident?

Reply:
Kind of, it's meant to be a placeholder of "fixme"; I found out whilst adding types that I'd missed the groupby matcher as we don't actually use it anywhere, but I need to re-add it along with the typing

Reply:
Fixed